uninstall.php

```diff
@@ -0,0 +1,156 @@
+	// in which case they cannot be queried for deletion.
+	if ( wp_using_ext_object_cache() ) {
+		return;
+	}
+
+	global $wpdb;
+
+	$transient_groups = [
+		'amp-parsed-stylesheet-v%',
+		'amp_error_index_counts',
+		'amp_img_%',
+		'amp_lock_%',
+		'amp_new_validation_error_urls_count',
+		'amp_plugin_activation_validation_errors',
+		'amp_themes_wporg',
+	];
+
+	$where_clause = [];
+
+	foreach ( $transient_groups as $transient_group ) {
+		if ( false !== strpos( $transient_group, '%' ) ) {
+			$where_clause[] = $wpdb->prepare(
+				' option_name LIKE %s OR option_name LIKE %s ',
+				"_transient_$transient_group",
+				"_transient_timeout_$transient_group"
+			);
+		} else {
+			$where_clause[] = $wpdb->prepare(
+				' option_name = %s OR option_name = %s ',
+				"_transient_$transient_group",
+				"_transient_timeout_$transient_group"
+			);
+		}
+	}
+
+	// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching -- Cannot cache result since we're deleting the records.
+	$wpdb->query(
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared -- See use of prepare in foreach loop above.
+		"DELETE FROM $wpdb->options WHERE " . implode( ' OR ', $where_clause )
+	);
+}
+
+/**
+ * Delete AMP Validated URL posts.
+ *
+ * @return void
+ */
+function delete_posts() {
+
+	global $wpdb;
+
+	$current_page = 1;
+	$per_page     = 1000;
+	$post_type    = 'amp_validated_url';
+
+	do {
+		$offset = $per_page * ( $current_page - 1 );
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching -- Cannot cache result since we're deleting the records.
+		$post_ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT ID FROM $wpdb->posts WHERE post_type = %s LIMIT %d OFFSET %d;",
+				$post_type,
+				$per_page,
+				$offset
+			)
+		);
+
+		if ( empty( $post_ids ) || ! is_array( $post_ids ) ) {
+			break;
+		}
+
+		foreach ( $post_ids as $post_id ) {
+			wp_delete_post( $post_id );
+		}
+
+		$current_page++;
+	} while ( ! empty( $result ) );
+}
+
+/**
+ * Delete AMP validation error terms.
+ *
+ * @return void
+ */
+function delete_terms() {
+
+	global $wpdb;
+
+	$current_page = 1;
+	$per_page     = 1000;
+	$taxonomy     = 'amp_validation_error';
+
+	do {
+		$offset = $per_page * ( $current_page - 1 );
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching -- Cannot cache result since we're deleting the records.
+		$term_ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT term_id FROM $wpdb->term_taxonomy WHERE taxonomy = %s LIMIT %d OFFSET %d;",
+				$taxonomy,
+				$per_page,
+				$offset
+			)
+		);
+
+		if ( empty( $term_ids ) || ! is_array( $term_ids ) ) {
+			break;
+		}
+
+		foreach ( $term_ids as $term_id ) {
+			wp_delete_term( $term_id, $taxonomy );
+		}
+
+		$current_page++;
+	} while ( ! empty( $result ) );
+}
+
+delete_options();
+delete_posts();
+delete_terms();
+delete_transients();
```

Comment:
There's one last thing to consider here, and that is _multisite_. When the plugin is uninstalled on a multisite network, the data will only get purged from the main site and not the other sites.

We may need do do what was done in `BackgroundTaskDeactivator` to loop over all sites in the network if it is `! wp_is_large_network()` and then `switch_to_blog()` for each and run these functions. On the other hand, if `wp_is_large_network()` is true, then what's the solution? On one hand, we could do something similar to `BackgroundTaskDeactivator` and warn the user that deactivating and uninstalling will leave data behind:

https://github.com/ampproject/amp-wp/blob/819226ec9f096b10cb11c542a068e542e9cd5fee/src/BackgroundTask/BackgroundTaskDeactivator.php#L134-L160

But then how would the owner of a large site delete the old data? Should these functions be put into a separate include file like `includes/uninstall-functions.php` which `uninstall.php` can then `require` and use? For large multisite networks, should there be a new WP-CLI command to reset the data which also calls these functions? If so, then a large network owner could use WP-CLI to walk over each site to purge the data like as follows:

```bash
for url in $(wp site list --field=url); do
	if wp plugin is-active amp; then
		wp amp reset-data --yes
		wp plugin deactivate amp
	fi
done
wp plugin deactivate --network 
wp plugin uninstall amp
```

Reply:
cf. how Site Kit does this: 

https://github.com/google/site-kit-wp/blob/develop/uninstall.php
https://github.com/google/site-kit-wp/blob/d6ea05554e20353df9e770abd193c98b006cdf79/includes/Core/Util/Uninstallation.php

It doesn't seem the Site Kit is handling multisite uninstall currently either.

What then may make sense to do is to create a new `Uninstallation` service which:

1. Has these functions as methods.
2. Provides one method to `reset_data()` which calls these 4 methods.
3. Registers a WP-CLI command to `reset-data`.

The `uninstall.php` can then load up `amp.php` or invoke the necessary logic so that `amp_bootstrap_plugin()` is called, and then the service would be registered and ready to go. For example:

```php
<?php
/**
 * Plugin uninstall file.
 *
 * @package AMP
 */

namespace AmpProject\AmpWP;

// If uninstall.php is not called by WordPress, then die.
if ( ! defined( 'WP_UNINSTALL_PLUGIN' ) ) {
	die;
}

require_once __DIR__ . '/amp.php';
amp_bootstrap_plugin();

$uninstallation_service =  Services::get( 'uninstallation' );

$uninstallation_service->reset_data();

if ( is_multisite() && ! wp_is_large_network() ) {
	foreach ( get_sites( [ 'fields' => 'ids' ] ) as $site_id ) {
		switch_to_blog( $site_id );
		$uninstallation_service->reset_data();
		restore_current_blog();
	}
}
```

Thoughts?

Reply:
Oh, and Web Stories _is_ multisite-aware and they're using the `/includes/uninstall.php` approach: https://github.com/google/web-stories-wp/blob/main/uninstall.php

Nevertheless, it's not accounting for large networks. Perhaps that would be good for us to do to begin with as well. A site owner could always just do `wp plugin uninstall amp`  from the command line on a large network, and that will prevent timeout problems. Then we can avoid having to create new a new WP-CLI command.

So let's adopt the Web Stories model.